### DISCUSSION TITLE
Direct2D DRAW Backend

### DISCUSSION THREAD

**Discussion by qtxie** [red/red#4764] [Sandbox]
We're trying to replace the current GDI/GDI+ DRAW backend with Direct2D on Windows. 
Source code: https://github.com/red/red/tree/Direct2D
Prebuilt binary for testing: 
 ~~[gui-console-d2d.zip](https://github.com/red/red/files/5687415/gui-console-d2d.zip)~~
 [gui-console-d2d-6256a4771b46c974cd1b51069bc00260a83443ee.zip](https://github.com/red/red/files/5720644/gui-console-d2d-6256a4771b46c974cd1b51069bc00260a83443ee.zip)

Report any 🐛 in here are very welcome. 😎 

_NOTE:_ Minimum supported client: Platform Update for Windows 7.

**Comment by planetsizecpu** [comment]
I tested the prebuilt binary with my game and works as normal, althought I don't use draw.

**Comment by pekr** [comment]
Here goes, just a very bried test:

- opens black console behind the gui window
- Particles dome is faster by some 7-10 fps, but it draws correctly only one particle, the rest are some kind of gradient balls

**Reply by qtxie** [reply]
> opens black console behind the gui window

It's normal as it's a debug version.

> but it draws correctly only one particle

There is an updated version in D2D branch. https://github.com/red/code/blob/D2D/Scripts/particles.red

**Comment by hiiamboris** [comment]
Here's how D2D compares to master on W10:
![](https://i.gyazo.com/77a61282e80ee96bc66174c4a616bc7d.png)
As you can see there are 6 regressions. #4554 can be ignored I think, probably a fluke caused by me using debug build of D2D and release build of master. I'll add more info when I get there.

**Reply by hiiamboris** [reply]
#3847 failed because `base` of unspecified color draws text using ClearType, but when color is specified it uses normal antialiasing (probably not something we will want to fix). But it highlighted another problem: default font is much smaller on D2D branch:
![](https://i.gyazo.com/2c5895f20c094d4a73c47c1f73466314.png)
using code:
```
		l1: layout [backdrop white size 200x200]
		f1: make face! [
			type: 'base
			offset: 0x0
			size: 160x160
			color: white
			draw: [text 0x0 "Hi there"]
		]
		append l1/pane f1
```

**Reply by hiiamboris** [reply]
In #3789 gradients do not work at all, curves only work sometimes, and curvature is different from master. Example comparison, on the `diff` slide you can see that curvature changed:
![](https://i.gyazo.com/50280c07900e992f0c5eaabd5e5bf76d.png)
The white short line in the upper center of 'D' shape hints that color of the left slide is halfway between gradient start and end points, so maybe gradient scale is either too huge or too small (less than a pixel)?

**Reply by hiiamboris** [reply]
#3725 was triggered because the offset of text changed:
![](https://i.gyazo.com/33ea7b513b2bd09104807a42e09d28d3.png)
Using code:
```
font1: make font! [name: system/view/fonts/fixed size: 30 color: black]
view [
	backdrop white
	below
	box 120x40 "A B C" left top font font1
	box 120x40 draw [font font1 text 0x0 "A B C"]
]
```
Are we still using GDI+ for `box "text"` thing? Can we change it to D2D?

**Reply by hiiamboris** [reply]
#3564 plainly regressed. Using code:
```
view/no-wait [base 100x100 draw x: copy y: [box 10x0 20x10 box 20x0 10x10]]
probe x = y
?? x ?? y
```
I'm getting:
```
false
x: [box 10x0 20x10 box 10x0 20x10]
y: [box 10x0 20x10 box 20x0 10x10]
```

**Reply by hiiamboris** [reply]
#3180 when unspecified, uses black font color instead of OS default:
![](https://i.gyazo.com/70152b807f824146ac6b6ccf618e642a.png)
using code:
```
view [
	backdrop white
	base white     "split-^/text"
	box  #FFFFFF01 "split-^/text"
]
```

**Reply by hiiamboris** [reply]
#4006 regression test tests for other thing, but it caught regression in `push` - it doesn't restore the pen:
![](https://i.gyazo.com/a31a09bc76a924b0261ef9ff61d0fa3a.png)
Using code:
```
view [
	    box 400x400 coal draw [
	    	line-width 10
	        pen red
	        fill-pen green
	        box 10x10 300x300
	        push [
	            pen off
	            fill-pen blue
	            box 30x30 320x320
	        ]
	        box 50x50 340x340
	    ]
]
```
Ignoring the subtle intensity differences (254 vs 255), but the last box should be green.

**Reply by hiiamboris** [reply]
#4356 regressed too a little bit - it eats a part of the last line:
![](https://i.gyazo.com/b9487610879fd660afa6b16d511fbc25.png)
Using code: `view [button 80 wrap "█^/█^/█^/█^/█^/█^/█^/█^/█^/█"]`

**Comment by greggirwin** [comment]
As always, the details, graphs, and example pics are enormously helpful @hiiamboris.

**Comment by qtxie** [comment]
@hiiamboris Thanks a lot. 👍 Regressions, even new bugs are expected.

> Are we still using GDI+ for box "text" thing? Can we change it to D2D?

Yes. We can change it to D2D. But it has some overheads. Creating the DX swapchain and buffer to just draw a few texts is expensive. Even worse, if it's a box face (layered window), we have to copy the content in the swapchain into the DC. Using GDI+ we just need to draw on the existing DC.

**Comment by qtxie** [comment]
In #3789, fixed the curve issue. The gradient issue is caused by `GetBounds` returns an infinite big bound. I'm not sure if I misused the API or it's the correct bound.

**Reply by hiiamboris** [reply]
I wanted to rant here, but turns out it was worth a ticket: https://github.com/red/red/issues/4769
However notice that the curve is **not closed** on D2D. There were complaints that `shape` is closed without `close` keyword and that one must use `move 0x0` workaround, but since other platforms close it, so should D2D.

**Reply by greggirwin** [reply]
The `close` discussion sounds familiar.

**Reply by qtxie** [reply]
I think it's right on D2D. We have a `close` keyword in draw.
```
view [base draw [shape [move 5x5 line 50x20 line 10x60 close]]] 
```

**Reply by hiiamboris** [reply]
Then all other backends should be fixed

**Comment by ne1uno** [comment]
I couldn't run the prebuilt on win7 on an old laptop, but runs ok win 10 slightly newer laptop
```
*** Runtime Error 98: assertion failed
*** in file: /D/EE/QW/red/modules/view/backends/windows/direct2d.reds
*** at line: 1679
***
***   stack: gui/create-render-target 00000062h 005101D8h
***   stack: gui/create-render-target 01310A90h 021A6028h
***   stack: gui/get-hwnd-render-target 01310A90h false
***   stack: gui/OS-text-box-layout 0213D774h 00000000h 0 false
***   stack: ctx||469~text-box-metrics 0213D774h 0213D784h 3
***   stack: size-text
***   stack: ctx||631~update-cfg 00340FBCh
***   stack: ctx||600~apply-cfg 003412DCh
***   stack: ctx||600~load-cfg 003412DCh
***   stack: ctx||600~launch 003412DCh

w/prebuilt L512 i5  win7-64 onboard intel GPU probably openGL 2?
E530 i7 win10 onboard intel GPU probably openGL 3 runs ok
both old lenovo laptops >6 years old
```

**Reply by qtxie** [reply]
Platform Update for Windows 7 is required. https://www.microsoft.com/en-us/download/details.aspx?id=36805

**Reply by ne1uno** [reply]
thanks for looking, that platform update says it's already installed.
I already updated graphic drivers hoping to get a newer openGL.
DxDiag Version: 6.01.7601.17514 64bit Unicode  DirectX 11

**Reply by qtxie** [reply]
@ne1uno A flag was wrong in view backend. Fixed in commit 1f4879c603048b314cd244a7d2cd0bf2a6f39aeb.

https://github.com/red/red/files/5720644/gui-console-d2d-6256a4771b46c974cd1b51069bc00260a83443ee.zip

**Reply by ne1uno** [reply]
that works win7 now, gui console starts and most programs work so far.
Red 0.6.4 for Windows built 20-Dec-2020/7:10:38

**Comment by hiiamboris** [comment]
Some pen madness here:
```
cmds: [
    pen cyan fill-pen magenta line-width 4
    ; clip 0x0 200x200 [
        push [
            pen pattern 4x4 [line-width 0 fill-pen white box 0x0 2x2 box 2x2 4x4] 
            line-width 4
            box 0x0 200x20
        ]
        box 0x60 50x80
    ; ]
    box 0x30 100x50
]

view [base 300x200 draw cmds]
```
Yields expected result:
![](https://i.gyazo.com/29ca6e3e2bafd974d97ec5b6ccb13397.png)

Enable `clip` and the 2nd box disappears:
```
cmds: [
    pen cyan fill-pen magenta line-width 4
    clip 0x0 200x200 [
        push [
            pen pattern 4x4 [line-width 0 fill-pen white box 0x0 2x2 box 2x2 4x4] 
            line-width 4
            box 0x0 200x20
        ]
        box 0x60 50x80
    ]
    box 0x30 100x50
]

view [base 300x200 draw cmds]
```
![](https://i.gyazo.com/4d8b984b1ac10b864dbe4bf07870be4e.png)

Remove duplicated line-width and all lines disappear:
```
cmds: [
    pen cyan fill-pen magenta line-width 4
    clip 0x0 200x200 [
        push [
            pen pattern 4x4 [line-width 0 fill-pen white box 0x0 2x2 box 2x2 4x4] 
            box 0x0 200x20
        ]
        box 0x60 50x80
    ]
    box 0x30 100x50
]

view [base 300x200 draw cmds]
```
![](https://i.gyazo.com/671739265441106caa2096d70c3eda56.png)
